File: core/src/main/java/javabot/javadoc/JavadocClass.java
package javabot.javadoc;

import java.util.ArrayList;
import java.util.List;

import com.antwerkz.maven.SPI;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.annotations.Index;
import org.mongodb.morphia.annotations.Indexes;
import org.mongodb.morphia.annotations.PrePersist;
import javabot.model.Persistent;
import org.bson.types.ObjectId;

@Entity(value = "classes", noClassnameStored = true)
@SPI(Persistent.class)
@Indexes({
    @Index("apiId"),
    @Index(value = "upperName"),
    @Index(value = "upperPackageName, upperName"),
    @Index(value = "apiId, upperPackageName, upperName"),
})
public class JavadocClass extends JavadocElement {
  @Id
  private ObjectId id;

  private String packageName;
  private String upperPackageName;
  private String name;
  private String upperName;
  private ObjectId superClassId;
  private List<JavadocMethod> methods = new ArrayList<>();
  private List<JavadocField> fields = new ArrayList<>();

  public JavadocClass() {
  }

  public JavadocClass(final JavadocApi api, final String pkg, final String name) {
    packageName = pkg;
    this.name = name;
    setApiId(api.getId());
    setDirectUrl(api.getBaseUrl() + pkg.replace('.', '/') + "/" + name + ".html");
    setLongUrl(api.getBaseUrl() + "index.html?" + pkg.replace('.', '/') + "/" + name + ".html");
  }

  public ObjectId getId() {
    return id;
  }

  public void setId(final ObjectId classId) {
    id = classId;
  }

  public String getPackageName() {
    return packageName;
  }

  public void setPackageName(final String name) {
    packageName = name;
  }

  public String getName() {
    return name;
  }

  public void setName(final String name) {
    this.name = name;
  }

  public ObjectId getSuperClassId() {
    return superClassId;
  }

  public void setSuperClassId(final ObjectId javadocClassId) {
    superClassId = javadocClassId;
  }

  public void setSuperClassId(final JavadocClass javadocClass) {
    superClassId = javadocClass.getId();
  }

  public String getUpperName() {
    return upperName;
  }

  public void setUpperName(final String upperName) {
    this.upperName = upperName;
  }

  public String getUpperPackageName() {
    return upperPackageName;
  }

  public void setUpperPackageName(final String upperPackageName) {
    this.upperPackageName = upperPackageName;
  }

  @PrePersist
  public void uppers() {
    upperName = name.toUpperCase();
    upperPackageName = packageName.toUpperCase();
  }

  @Override
  public String toString() {
    return packageName + "." + name;
  }
}

File: src/main/kotlin/javabot/javadoc/JavadocClass.kt
package javabot.javadoc

import org.mongodb.morphia.annotations.Entity
import org.mongodb.morphia.annotations.Field
import org.mongodb.morphia.annotations.Index
import org.mongodb.morphia.annotations.Indexes
import org.mongodb.morphia.annotations.PrePersist

@Entity(value = "classes", noClassnameStored = true)
@Indexes(Index(fields = arrayOf(Field("apiId"))), Index(fields = arrayOf(Field("upperName"))),
      Index(fields = arrayOf(Field("upperPackageName"), Field("upperName") )),
      Index(fields = arrayOf(Field("apiId"), Field("upperPackageName"), Field("upperName") )))
open class JavadocClass : JavadocElement {
    lateinit var packageName: String
    lateinit var name: String
    lateinit var fqcn: String
    var parentClass: String? = null
    var interfaces = mutableListOf<String>()
    var isClass = false
    var isEnum = false
    var isInterface = false
    var isAnnotation = false
    var typeVariables = mutableListOf<String>()

    lateinit var upperPackageName: String
    lateinit var upperName: String

    constructor() {
    }

    constructor(api: JavadocApi, pkg: String, name: String) {
        packageName = pkg
        this.name = name
        fqcn = "$packageName.$name"
        apiId = api.id
        url = api.baseUrl + "index.html?" + fqcn.replace('.', '/') + ".html"
    }

    @PrePersist
    fun uppers() {
        upperName = name.toUpperCase()
        upperPackageName = packageName.toUpperCase()
    }

    override fun toString(): String {
        return packageName + "." + name
    }
}

enum class Visibility {
    PackagePrivate,
    Private,
    Protected,
    Public
}
